spec/haskell/src/SEL4/Object/SchedContext.lhs

```diff
@@ -256,96 +256,79 @@ This module uses the C preprocessor to select a target architecture.
 >         maybeAddEmptyTail scPtr
 >       else do
 >         let unused = newBudget - rAmount head
->         let new = Refill { rTime = rTime head + newPeriod - unused, rAmount = unused }
->         scheduleUsed scPtr new
+>         let new = Refill { rTime = rTime head + newPeriod, rAmount = unused }
+>         refillAddTail scPtr new
 
 > scheduleUsed :: PPtr SchedContext -> Refill -> Kernel ()
 > scheduleUsed scPtr new = do
 >     sc <- getSchedContext scPtr
 >     assert (scRefillMax sc > 0) "scPtr should be active"
 >     empty <- refillEmpty scPtr
 >     if empty
```

Comment:
I definitely did want a branch in this function for the empty case, but I forget the exactly reason. I'll ask Curtis whether he is happy to have a branch in the C for this too. We should never call this function where the sc has empty refills (it's now called only in `refill_budget_check`, by the way), but if it makes verification easier, then I would think it would be better to have the branch there

Reply:
I asked Curtis, and he said that if the branch if not required, then he will have a hard time convincing Anna to keep it there. I would think that we would very much like to define `schedule_used` to be a total function, so I can't see how we could get around having the branch in the abstract spec. And in the C, in order to properly access the `REFILL_TAIL`, the sc must not be empty. So I can't see a good way around this. Should an assert be added to the C? Or at least a comment saying that it shouldn't be called on an empty list?

Reply:
OK so Curtis still does not want a branch to be added to the C, but he is happy to move the assert that the refill list is nonempty to the beginning of `schedule_used`. So I think the best approach would be to leave the abstract and Haskell as they are. When we prove refinement, we can assume that the refill list is nonempty (we would call `refill_budget_check` only on a scheduling context that satisfies `valid_refills`, and so in particular is not empty). Does that sounds OK?

Reply:
Yes, that sounds great. Thanks.